List and detail view

Ganapathy-dev opened the pull request:
# **Pull Request: Add List and Detail Views for Restaurants**

This PR implements the following features:

### **List View**
- Adds `RestaurantListView` with:
  - Filtering by city, food type (veg/non-veg), cuisines, and open/closed status.
  - Sorting by rating and cost for two (high-to-low and low-to-high).

### **Detail View**
- Adds `RestaurantDetailView` with:
  - Detailed restaurant information, including categorized dishes (veg, non-veg, vegan).
  - User interaction features to:
    - Bookmark or unbookmark a restaurant (`ToggleBookmarkView`).
    - Mark a restaurant as visited or unvisited (`ToggleVisitedView`).
    - Submit reviews and ratings directly from the detail page if user is authenticated.

## **Testing**
- Added test cases for:
  - Filtering and sorting in `RestaurantListView`.
  - Rendering and context data for `RestaurantDetailView`.

Ganapathy-dev committed:
Implement List View

-create view to list all available restaurant
-it should be able to sort by cost, rating (high to low vice versa)
-it should be filtered by city, cuisine, food, rating, cost

Ganapathy-dev committed:
Restaurant List and Detail view

-create Restaurant list view with filter and sorting functinality
-create Detail view, shows reviews and Menu dishes allows user to
bookmark and mark visited if login

shantanutomarr commented on restaurant/views.py:
```diff
@@ -1,3 +1,147 @@
+            self.sort_by=self.form.cleaned_data.get('sort_by')
+        
+    def filterqueryset(self):
+            if self.city:
+                self.queryset=self.queryset.filter(location__icontains=self.city)
+
+            if self.food_type:
+                query=Q()
+                for type in self.food_type:
+                     query |= Q(food_type__iregex=rf'(^|,){type}(,|$)')
+                self.queryset=self.queryset.filter(query)
+
+            if self.cuisines:
+                query=Q()
+                for cuisine in self.cuisines:
+                    query |=Q(cuisines__icontains=cuisine)
+                self.queryset=self.queryset.filter(query)
+            
+            if self.rating:
+                self.queryset=self.queryset.filter(rating__gte=self.rating)
+
+            if self.cost_of_two:
+                self.queryset=self.queryset.filter(cost_of_two__gte=self.cost_of_two)
+
+            if self.is_open:
+                self.queryset=[restaurant for restaurant in self.queryset if restaurant.is_open()]
+
+        
+    def sortqueryset(self):
+                sort_order=self.form.cleaned_data.get('sort_order')
+                if sort_order:
+                    if sort_order=='high to low':
+                        self.queryset=self.queryset.order_by(f'-{self.sort_by}')
+                    if sort_order=='low to high':
+                        self.queryset=self.queryset.order_by(self.sort_by)
+                else:
+                    self.queryset=self.queryset.order_by('-created_at')
+
```
Use django_filters package for filtering and sorting. https://pypi.org/project/django-filter/

Ganapathy-dev replied:
I already tired using django-filter it not properly working for few fields they are : 
food type (MultiselectField), 
cuisine (MultiselectField), and 
is_open (Method in models) so I implemented this way.

Ganapathy-dev committed:
Implement django-filter

-Refactor filter and sorting in view, include djang-filter for
filtering and sorting

Ganapathy-dev committed:
Resolved failing test cases
